enhancements/cluster-runtime-constraints.md

```diff
@@ -0,0 +1,159 @@
+
+## Summary
+
+This proposal presents the specification and usage to enable runtime constraints to be considered during operator dependency resolution. This feature will ensure Operator Lifecycle Manage (OLM) to install operators that satisfy dependency requirements and runtime constraints if present.
+
+## Motivation
+
+At the moment, OLM resolves a set of operators which seem that they will work together based on dependency requirements that are specified by operator bundles. However, it is possible that those resolved operators will not work on the clusters due to cluster/runtime constraints. For example, if the cluster is running a specific Kubernetes version that is not compatible with the operators, the installed operators may fail to work properly on that given cluster.
+
+The cluster runtime constraints are often unknown by the operator authors and should be specified by cluster admins. However, currently, there is no mechanism for cluster admins to specify runtime constraints that OLM can understand and take under consideration during installation.
+
+### Goals
+- Provide mechanism (including specification and guideline) for cluster admins to specify cluster runtime constraints
+- Provide specification and guideline for author operators to specify runtime constraints/requirements if needed
+- Block the updates/installations of operators based on cluster runtime constraints
+- Allow the cluster runtime constraints to be generic and configurable by cluster admins
+
+### Non-Goals
+
+- Provide mechanism to retrieve/populate cluster runtime constraints automatically
+- Allow cluster runtime constraints to be scoped
+- Allow cluster runtime constraints to come from multiple/different sources
+- Allow optional cluster runtime constraints
+
+
+## Proposal
+
+To provide a generic solution for cluster runtime constraints, a library named [cel-go](https://github.com/google/cel-go) that provides expression evaluation capabilities is utilized in this proposal.
+
+### User Stories
+
+#### Constraints specified by cluster admin
+
+As a cluster admin, I would like to specify a list of cluster runtime constraints that are associated with a given cluster and all installable operators will meet these requirements.
+
+#### Constraint properties specified by operator author
+
+As an operator author, I would like to provide a list of runtime properties that may be needed to be considered before installation to ensure the operator working properly on the cluster.
+
+### Implementation Details
+
+#### Cluster Runtime Constraint Configmap
+
+Cluster admins can specified cluster runtime constraints in a configmap (named `olm-runtime-constraints`) in `olm` namespace using this format:
+
+```yaml=
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: olm-runtime-constraints
+  namespace: olm
+data:
+    properties: <list-of-expression-constraints-in-json>
+```
+
+Each constraint is specified using the following syntax:
```

Comment:
Why are we calling the field `properties` when the value is a list of constraints?

Do we allow _any_ properties in this list, where constraints are just a type of property? If so, it would be good to clarify that.

Reply:
Everything is properties in the resolver. So constraint is just another type of properties as you said. And yes you can specify properties in this configmap.